Port `sort` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62391

Tracking issue: #55070

Differential Revision: [D30903992](https://our.internmc.facebook.com/intern/diff/D30903992)

ysiraichi committed:
Port `sort` kernel to structured kernels.

Tracking issue: #55070

This PR also makes `sort.stable` kernel changes the `optional<bool> stable` parameter to a
required `bool stable` one. The reason being that both CPU and CUDA kernels called
`TORCH_INTERNAL_ASSERT(stable.has_value())` (i.e. can't see why it was optional).

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

This PR also makes `sort.stable` kernel changes the `optional<bool> stable` parameter to a
required `bool stable` one. The reason being that both CPU and CUDA kernels called
`TORCH_INTERNAL_ASSERT(stable.has_value())` (i.e. can't see why it was optional).

[ghstack-poisoned]

ysiraichi commented:
While porting this kernel, I got into something I'm not sure we can address without changing kernel behavior and performance. Apparently, when there are no outputs defined, sort allocates them in different ways on CUDA:

1. If the input is small enough and short is unstable, the output will be contiguous
2. Else, the output will be strided the same as the input ([there is a test for this](https://github.com/pytorch/pytorch/blob/fd8004b42e2a2348ec8837e3fb524b960c1b4cdb/test/test_sort_and_select.py#L184-L189), specifically)

Clearly, changing where the output kernels get allocated are likely to affect the kernel performance. So, the question is: **how to allocate output tensors for `sort`?** In this porting, I choose to allocate tensors with the same strides as the input ([there is a comment in this test](https://github.com/pytorch/pytorch/blob/fd8004b42e2a2348ec8837e3fb524b960c1b4cdb/test/test_sort_and_select.py#L186-L187) saying it should be the ideal behavior). However, there are some CI failures (reason why I haven't opened it yet). The reason being a bug in the implementation of the kernel for small tensors (see https://github.com/pytorch/pytorch/issues/62645).

Here are some other options that I can think of:

1. Leave it unstructured, so that we keep the expected perfomance of the functional variant
2. Make it always return outputs that are strided like the input
3. Implement an alternative unstructured version of this kernel, so that people that don't really need meta support can take advantage of that

@ezyang @bdhirsh  What do you think?

bdhirsh commented:
Yukio and I talked about this a little over slack - this seems like a pretty similar situation to https://github.com/pytorch/pytorch/pull/62513. Whereas `native_batch_norm` sometimes produces a different output dtype on cuda, `sort` sometimes produces different strides on cuda (both for performance reasons - if the sort is marked with stable=False, and # elements is < 2048).

It would be nice if we could make the meta function device agnostic

> Here are some other options that I can think of:
(1) Leave it unstructured, so that we keep the expected perfomance of the functional variant
(2) Make it always return outputs that are strided like the input
(3) Implement an alternative unstructured version of this kernel, so that people that don't really need meta support can take advantage of that

I'm not sure how viable (2) is, since it'll hurt perf (I tried to find the original PR that added that optimization, but it looks like this kernel was ported from TH).

(3) is kind of annoying because it means a new user API that we have to maintain (and what if we ever fix this problem in the future? We'd have to either maintain or deprecate it

I think it's reasonable to push off making this structured for now. Another approach that might be reasonable would be to add some logic in the meta function like:
```
if (device == kMeta && stable == false && size <= 2048) *throw an error*
```

That way we still have a valid meta function and we're able to do the port, but people trying to use the meta tensor API with `torch.sort()` will only hit problems in the case where their inputs would create a different meta output depending on the backend. It's still not great, but it allows you to land your PR and also doesn't seem like regretful work.

ezyang commented:
Fixing the behavior of CUDA to preserve strides in all conditions seems like the right thing to do to me. You can still recover the old performance by explicitly passing a tensor with the old style strides.

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ezyang commented:
@bdhirsh this one for you?

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7289,27 +7289,24 @@
   dispatch:
-    CPU: sort_out_cpu
-    CUDA: sort_out_cuda
+    CPU, CUDA: sort_out
```
Technically `sort_out` and `sort` are both composite now, right? Since you implemented them as re-dispatching into the `sort.stable` overload. So you could mark them as `CompositeExplicitAutograd` instead.

ysiraichi replied:
You are right. I will do it.

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented:
@bdhirsh All done here. Do you have some time to take a look at this PR?

bdhirsh commented:
just came back from PTO. Ready to import!

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30903992).

bdhirsh commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -13,30 +13,42 @@
+      "sort(): c10::optional<bool> for stable has to have value.");
+  maybe_wrap_dim(dim, self.dim());
+  const auto& values = maybe_get_output(0);
+  const auto& indices = maybe_get_output(1);
```
Looks like `values` and `indices` are unused? (internal lint tests are failing on this)

ysiraichi replied:
You are right. I will fix this one.

ysiraichi committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D30903992](https://our.internmc.facebook.com/intern/diff/D30903992)

[ghstack-poisoned]

ysiraichi commented:
@bdhirsh I've removed those! Could you take a look?

bdhirsh reviewed (approved):
lgtm!

bdhirsh committed:
Update on "Port `sort` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D30903992](https://our.internmc.facebook.com/intern/diff/D30903992)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30903992).
